test/fakeserver.py

```diff
@@ -0,0 +1,200 @@
+import json
+import threading
+import time
+from http.server import HTTPServer, BaseHTTPRequestHandler
+from itertools import repeat
+from typing import Iterable
+
+# How long to wait in seconds before sending a response in a test
+# case where we are checking that we will wait forever.
+FOREVER_BLOCKING_TIME = 10
+
+
+class BlockingServer(HTTPServer):
+
+    # Amount of time to wait, in seconds, before sending a GET response.
+    get_blocking = repeat(StopIteration, 1)
+
+    # Has the server been _stopped so it no longer needs to
+    # return responses.
+    _stopped = False
+    _lock = threading.Lock()
+
+    # noinspection PyPep8Naming
+    def __init__(self, server_address, RequestHandlerClass):
+        super().__init__(server_address, RequestHandlerClass)
+
+    def is_stopped(self) -> bool:
+        with self._lock:
+            return self._stopped
+
+    def shutdown(self) -> None:
+        super().shutdown()
+        with self._lock:
+            self._stopped = True
+
+
+class FakeHandler(BaseHTTPRequestHandler):
+    """Override the server type from BaseServer."""
+
+    def _server(self) -> BlockingServer:
+        """Casting self.server for Python checking."""
+        # noinspection PyTypeChecker
+        return self.server
+
+    def _block(self, delay_list: Iterable) -> None:
+        """Delay for a period of time.
+
+        Attributes:
+            delay_list: Top value will be removed and used as the number
+            of seconds to wait.
+        """
+        delay = delay_list.__next__()
+        if delay == StopIteration:
+            # noinspection PyProtectedMember
+            msg = f"delay_list is empty: {self._server()._testMethodName}"
+            raise Exception(msg)
+
+        while delay > 0 and not self._server().is_stopped():
```

Comment:
This assumes that the only source of delay is the sleep, which may not be true. Better to use `timer.monotic` and check/calculate a deadline against it.